psi4/src/psi4/fisapt/fisapt.cc

```diff
@@ -1580,12 +2600,457 @@ void FISAPT::exch() {
+//for the opposite-spin case, we have to pick a spin-up slice that is noncontiguous with x/y
+        std::vector<SharedMatrix> hold_temp;
+    
+        hold_temp.push_back(Tbig->get_block(Slice(zero, nadim), Slice(nadim, naaxdim)));
+        hold_temp.push_back(Tbig->get_block(Slice(naadim, naaxdim), Slice(nadim, naaxdim)));
+        auto Dii_os = linalg::vertcat(hold_temp);
+        hold_temp.clear();
+    
+        hold_temp.push_back(Tbig->get_block(Slice(zero, nadim), Slice(naaxbdim, naaxbbydim)));
+        hold_temp.push_back(Tbig->get_block(Slice(naadim, naaxdim), Slice(naaxbdim, naaxbbydim)));
+        auto Dij_os = linalg::vertcat(hold_temp);
+        hold_temp.clear();
+    
+        hold_temp.push_back(Tbig->get_block(Slice(naaxdim, naaxbdim), Slice(nadim, naaxdim)));
+        hold_temp.push_back(Tbig->get_block(Slice(naaxbbdim, naaxbbydim), Slice(nadim, naaxdim)));
+        auto Dji_os = linalg::vertcat(hold_temp);
+        hold_temp.clear();
+
+        hold_temp.push_back(Tbig->get_block(Slice(naaxdim, naaxbdim), Slice(naaxbdim, naaxbbydim)));
+        hold_temp.push_back(Tbig->get_block(Slice(naaxbbdim, naaxbbydim), Slice(naaxbdim, naaxbbydim)));
+        auto Djj_os = linalg::vertcat(hold_temp);
+        hold_temp.clear();
+
+//half-transformed matrices for the JK operators
+        auto Dsj_ssh = linalg::doublet(matrices_["AlloccA"], Dij_ss, false, false);
+        Dsj_ssh->add(linalg::doublet(matrices_["AlloccB"], Djj_ss, false, false));
+        auto Dsj_osh = linalg::doublet(matrices_["AlloccA"], Dij_os, false, false);
+        Dsj_osh->add(linalg::doublet(matrices_["AlloccB"], Djj_os, false, false));
+
+        Cl.clear();
+        Cr.clear();
+        Cl.push_back(Dsj_ssh);
+        Cr.push_back(matrices_["AlloccB"]);
+        Cl.push_back(Dsj_osh);
+        Cr.push_back(matrices_["AlloccB"]);
+    
+        jk_->compute();
+
+        std::shared_ptr<Matrix> J_ss = J[0];
+        std::shared_ptr<Matrix> K_ss = K[0];
+        std::shared_ptr<Matrix> J_os = J[1];
+        std::shared_ptr<Matrix> K_os = K[1];
+
+        K_ss->transpose_this();
+        K_os->transpose_this();
+
+//finish the transformation of the D matrices
+        auto Dsj_ss = linalg::doublet(Dsj_ssh, matrices_["AlloccB"], false, true);
+        auto Dsj_os = linalg::doublet(Dsj_osh, matrices_["AlloccB"], false, true);
+        auto Dri_ss = linalg::triplet(matrices_["AlloccA"], Dii_ss, matrices_["AlloccA"], false, false, true);
+        Dri_ss->add(linalg::triplet(matrices_["AlloccB"], Dji_ss, matrices_["AlloccA"], false, false, true));
+        auto Dri_os = linalg::triplet(matrices_["AlloccA"], Dii_os, matrices_["AlloccA"], false, false, true);
+        Dri_os->add(linalg::triplet(matrices_["AlloccB"], Dji_os, matrices_["AlloccA"], false, false, true));
+    
+        double Enuc = 0.0;
+        double** Enuc2p = matrices_["E NUC"]->pointer();
+        Enuc += 2.0 * Enuc2p[0][1];  // A - B
+
+//The full first-order energy (elst+exch) is given by Eq. (21)
```

Comment:
Another small cleanup thing, could be adjusting comments like these so that they are aligned with they code they are commenting on.

Reply:
I opted against shifting all comments far to the right, but I'll be happy to do so if that's the generally preferred style. However, I was able to remove some unnecessary indentation so this should now be less of an issue.

Reply:
Someday a formatter will have its way with the comments. Then we can decide if it's worth resisting. :-) Thanks for the indentation change.